### DISCUSSION TITLE
about locks

### DISCUSSION THREAD

**Discussion by ryan053857** [roadrunner-server/roadrunner#2039] [Q&A]
I'm new to roadrunner, and I'm a little confused about Locks. The documentation mentions that Locks are memory-based, which means that if I deploy multiple roadrunner instances, Locks will fail. Then it's not called a distributed lock.

In fact, my current application is not too complicated, and the concurrency is not very large. But I plan to deploy 2 instances on one machine with docker swarm, so that rolling updates can be achieved without stopping when updating the service. Obviously Locks won't work in this scenario.

What are the Locks best practices? Are you considering implementing distributed Locks by adding redis or database as the back end of locks?

**Comment by rustatian** [comment]
Hey @shellphy 👋 
Locks are distributed, since you can use 1 RR instance to control locks from the different distribute applications.
Locks are not intended to solve problems with green-blue deploy. Locks are mutexes to solve the problem of concurrent use of some resource. For example, writing and reading from file from multiply processes.

**Comment by ryan053857** [comment]
Are there any plans to develop a distributed lock based on redis or database? It feels like a public need, too

**Reply by rustatian** [reply]
We have a Redis KV plugin, feel free to use it 😃
